include/alarm_heap.hpp

```diff
@@ -0,0 +1,198 @@
+  void update_alarm_in_heap(AlarmDef::Severity new_severity,
+                            uint64_t time_to_delay);
+
+  AlarmInHeap* alarm_in_heap() { return _alarm_in_heap; }
+  AlarmDef::Severity get_severity() { return _severity; }
+  void set_severity(AlarmTableDef& alarm_table_def);
+
+  // Determines whether this alarm should be resent after failure
+  bool should_resend_alarm(AlarmDef::Severity severity)
+  {
+    // Resend the alarm if its severity and the current severity match,
+    // and there isn't already an alarm in the heap waiting to update
+    // the severity.
+    return ((_severity == severity) && (!_alarm_in_heap->in_heap()));
+  }
+
+private:
+  AlarmInHeap* _alarm_in_heap;
+  AlarmDef::Severity _severity;
+};
+
+typedef std::map<unsigned int, AlarmInfo*> AllAlarmsStateMap;
+
+// This class accepts alarm triggers, checks that they correspond to valid
+// alarms, and passes valid alarms to the alarm trap sender to send as SNMP
+// INFORMs. It also filters the alarms.
+class AlarmHeap
+{
+public:
+  // Times to send an alarms. If we're reducing the severity of an alarm we
+  // add a delay of 30secs (to protect against flicker). Increased severity
+  // alarms or resyncing alarms are sent immediately.
+  const static uint64_t ALARM_REDUCED = 30000;
+  const static uint64_t ALARM_RETRY_DELAY = 30000;
+  const static uint64_t ALARM_RESYNC_DELAY = 0;
+  const static uint64_t ALARM_INCREASED = 0;
+
+  // Constructor/Destructor
+  AlarmHeap(AlarmTableDefs* alarm_table_defs);
+  virtual ~AlarmHeap();
+
+  // Generates an alarmActiveState inform if the identified alarm is not
+  // of CLEARED severity and not already active (or subject to filtering).
+  // Generates an alarmClearState inform if the identified alarm is of a
+  // CLEARED severity and an associated alarm is active (unless subject
+  // to filtering).
+  virtual void issue_alarm(const std::string& issuer,
+                   const std::string& identifier);
+
+  static void* heap_sender_function(void* data);
+
+  void heap_sender();
+
+  // Runs through all currently active alarms and adds them to the alarm heap
+  // to send immediately.
+  virtual void sync_alarms();
+
+  // Handles the case where an INFORM sent by the trap sender timed out.
+  // @params alarm_table_def - Definition of the alarm (index/severity) we
+  //                           tried to send.
+  void handle_failed_alarm(AlarmTableDef& alarm_table_def);
+
+  // Whether the program has terminated
+  volatile bool _terminated;
+
+private:
+  void update_alarm_in_heap(AlarmInfo* alarm_info,
+                            AlarmDef::Severity severity,
+                            uint64_t alarm_pop_time);
+
+  AlarmTableDefs* _alarm_table_defs;
+  AlarmTrapSender* _alarm_trap_sender;
+  TimerHeap _alarm_heap;
+
+  // Map holding all alarms. This maps the index of an alarm to information
+  // about the alarm (its current severity and its representation in the heap).
```

Comment:
Can we be a bit more precise than "current severity"? I think it's specifically "the severity we last tried to send to the NMS" (not "the severity we last successfully sent to the NMS" or "the severity we were last told about").